### ISSUE TITLE
Protocol Buffers v3.0.2: Cannot link GRPC with protobuf SHARED library

### ISSUE THREAD

**Issue by KindDragon** [google/protobuf#2219]
Issue with project `grpc_plugin_support`. Compiler MSVC2015

```
4>csharp_generator.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::csharp::GetFileNamespace(class google::protobuf::FileDescriptor const *)" (?GetFileNamespace@csharp@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVFileDescriptor@34@@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl grpc_csharp_generator::GetServices(class google::protobuf::FileDescriptor const *,bool,bool,bool)" (?GetServices@grpc_csharp_generator@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVFileDescriptor@protobuf@google@@_N11@Z)
4>csharp_generator.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::csharp::GetClassName(class google::protobuf::Descriptor const *)" (?GetClassName@csharp@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVDescriptor@34@@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl grpc_csharp_generator::`anonymous namespace'::GetMethodRequestParamMaybe(class google::protobuf::MethodDescriptor const *,bool)" (?GetMethodRequestParamMaybe@?A0x327ce4b9@grpc_csharp_generator@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVMethodDescriptor@protobuf@google@@_N@Z)
4>csharp_generator.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::csharp::GetReflectionClassName(class google::protobuf::FileDescriptor const *)" (?GetReflectionClassName@csharp@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVFileDescriptor@34@@Z) referenced in function "void __cdecl grpc_csharp_generator::`anonymous namespace'::GenerateServiceDescriptorProperty(class google::protobuf::io::Printer *,class google::protobuf::ServiceDescriptor const *)" (?GenerateServiceDescriptorProperty@?A0x327ce4b9@grpc_csharp_generator@@YAXPAVPrinter@io@protobuf@google@@PBVServiceDescriptor@56@@Z)
4>objective_c_generator.obj : error LNK2019: unresolved external symbol "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::objectivec::ClassName(class google::protobuf::Descriptor const *)" (?ClassName@objectivec@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVDescriptor@34@@Z) referenced in function "class std::map<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct std::less<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > > > __cdecl grpc_objective_c_generator::`anonymous namespace'::GetMethodVars(class google::protobuf::MethodDescriptor const *)" (?GetMethodVars@?A0xe130935c@grpc_objective_c_generator@@YA?AV?$map@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@U?$less@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@V?$allocator@U?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V12@@std@@@2@@std@@PBVMethodDescriptor@protobuf@google@@@Z)
4>D:\Work\grpc\.build\Debug\grpc_plugin_support.dll : fatal error LNK1120: 4 unresolved externals
```

https://github.com/grpc/grpc/issues/8259

**Comment by xfxyjwf** [comment]
It's recommended to static link protobuf library. If you are building protobuf library as DLLs, make sure define PROTOBUF_USE_DLLS when building your own code against protobuf lib.

**Comment by KindDragon** [comment]
Why protobuf_BUILD_SHARED_LIBS not define it automatically?

**Comment by KindDragon** [comment]
Aaa, you mean GRPC should define `protobuf_BUILD_SHARED_LIBS` when using protobuf

**Comment by KindDragon** [comment]
I tried, `PROTOBUF_USE_DLLS` doesn't help

**Comment by xfxyjwf** [comment]
I checked these methods, and it seems they are missing the proper LIBPROTOC_EXPORT annotations. Can you try if adding LIBPROTOC_EXPORT to https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/csharp/csharp_names.h (and the other objectivec one) helps?

**Comment by KindDragon** [comment]
This commit help with objective-c, but not with csharp https://github.com/KindDragon/protobuf/commit/90476bee10b80b95c070415fc8e8f0f931a8e78e

```
grpc_plugin_support.lib(csharp_generator.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::csharp::GetFileNamespace(class google::protobuf::FileDescriptor const *)" (__imp_?GetFileNamespace@csharp@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVFileDescriptor@34@@Z) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl grpc_csharp_generator::GetServices(class google::protobuf::FileDescriptor const *,bool,bool,bool)" (?GetServices@grpc_csharp_generator@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVFileDescriptor@protobuf@google@@_N11@Z) [D:\Work\vcpkg\buildtrees\grpc\x86-windows-rel\grpc_csharp_plugin.vcxproj]
grpc_plugin_support.lib(csharp_generator.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::csharp::GetClassName(class google::protobuf::Descriptor const *)" (__imp_?GetClassName@csharp@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVDescriptor@34@@Z) referenced in function "void __cdecl grpc_csharp_generator::`anonymous namespace'::GenerateClientStub(class google::protobuf::io::Printer *,class google::protobuf::ServiceDescriptor const *)" (?GenerateClientStub@?A0x0a9b87f7@grpc_csharp_generator@@YAXPAVPrinter@io@protobuf@google@@PBVServiceDescriptor@56@@Z) [D:\Work\vcpkg\buildtrees\grpc\x86-windows-rel\grpc_csharp_plugin.vcxproj]
grpc_plugin_support.lib(csharp_generator.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl google::protobuf::compiler::csharp::GetReflectionClassName(class google::protobuf::FileDescriptor const *)" (__imp_?GetReflectionClassName@csharp@compiler@protobuf@google@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PBVFileDescriptor@34@@Z) referenced in function "void __cdecl grpc_csharp_generator::`anonymous namespace'::GenerateServiceDescriptorProperty(class google::protobuf::io::Printer *,class google::protobuf::ServiceDescriptor const *)" (?GenerateServiceDescriptorProperty@?A0x0a9b87f7@grpc_csharp_generator@@YAXPAVPrinter@io@protobuf@google@@PBVServiceDescriptor@56@@Z) [D:\Work\vcpkg\buildtrees\grpc\x86-windows-rel\grpc_csharp_plugin.vcxproj]
```

**Comment by KindDragon** [comment]
This commit fix everything https://github.com/KindDragon/protobuf/commit/994722a0170d4f9a7f92435ce052755796af5967

**Comment by Kwizatz** [comment]
The problem is how protobuf exports symbols in Windows, it should mark specific member functions as declspec( export ) rather than the whole class as it is doing right now (see [here](http://stackoverflow.com/questions/16419318/one-way-of-eliminating-c4251-warning-when-using-stl-classes-in-the-dll-interface/35275889#35275889)).

I was once going to fix all of it, but it is just too much code, so I just changed what I needed locally and gave up on posting a pull request... since I am not the only one interested anymore, I'll see into resuming work on it.

For the impatient, What they have is this:

`declspec(export) class X
{
public:
X();
~X();
void foo();
private:
std::string bar;
}`

What they SHOULD HAVE is this:

`class X
{
public:
declspec(export) X();
declspec(export) ~X();
declspec(export) void foo();
private:
std::string bar;
}`

I am sure you can see how making the change can be extremely tedious.

**Comment by xfxyjwf** [comment]
Marking all methods with export annotation is not practical. If the problem is the use of STL classes  across dll boundaries, I think it's a better solution to avoid using them when we can.

**Comment by Kwizatz** [comment]
Agreed, and to be honest for my specific purposes I didn't have to change that many classes, so the problem may be marking more classes as LIBPROTOBUF_EXPORT than they need to be. Note that the problem I think is not related to return or argument types, but private member variables.

It doesn't matter whether the private dll stl class instance is being accessed or not, if you mark it as dllexport by marking the whole class, the linker will look for it but it will not find it unless you add hacks such as these [1](https://blogs.msdn.microsoft.com/sergey_babkins_blog/2015/09/08/exporting-c-classes-with-stl-members-from-dlls/) [2](https://social.msdn.microsoft.com/Forums/vstudio/en-US/df99b712-c00b-4af0-82fd-3764c8b6cbec/exporting-stdstring-from-a-dll-does-not-export-stdstringnpos?forum=vcgeneral), and those may bring in worse issues in the long run.

**Comment by KindDragon** [comment]
What wrong with changed from my PR #2227?

**Comment by xfxyjwf** [comment]
@KindDragon Your change looks good to me. @Kwizatz mentions there may be a problem with the use of std container type as member variables, but maybe it's not relevant in this particular case.

**Comment by Kwizatz** [comment]
Yes, sorry about my fuzz, I can just silence the warnings from protobuf in my project, there is one I cannot/wouldn't silence, but I already submited a pull request for that one 😅

**Comment by KindDragon** [comment]
Fixed by f39cf88a347767d25572f7344eab44359d872925
